lib/git.ml

```diff
@@ -1,10 +1,45 @@
 open Stdune
 open Sexplib.Conv
 
-let parse_ls_remote_line s =
-  match String.extract_blank_separated_words s with
-  | [ commit; ref ] -> Ok (commit, ref)
-  | _ -> Error (`Msg (Printf.sprintf "Invalid git ls-remote output line: %S" s))
+module Ls_remote = struct
+  let non_packed_suffix = "^{}"
+
+  let ref_arg ref = Bos.Cmd.(v ref % (ref ^ non_packed_suffix))
+
+  let parse_output_line s =
+    match String.extract_blank_separated_words s with
+    | [ commit; ref ] -> Ok (commit, ref)
+    | _ -> Error (`Msg (Printf.sprintf "Invalid git ls-remote output line: %S" s))
+
+  type search_result = { maybe_packed : string option; not_packed : string option }
+
+  let interpret_search_result sr =
+    match sr with
+    | { maybe_packed = None; not_packed = None } -> Error `No_such_ref
+    | { maybe_packed = Some commit; not_packed = None }
+    | { maybe_packed = _; not_packed = Some commit } ->
+        Ok commit
+
+  let commit_pointed_by ~ref output_lines =
+    let open Result.O in
+    let non_packed = ref ^ non_packed_suffix in
+    let is_ref full_ref = String.is_suffix ~suffix:ref full_ref in
+    let is_non_packed_ref full_ref = String.is_suffix ~suffix:non_packed full_ref in
+    let rec go acc lines =
+      match (acc, lines) with
+      | _, [] -> Ok acc
+      | { not_packed = None; _ }, (commit, full_ref) :: tl when is_non_packed_ref full_ref ->
+          go { acc with not_packed = Some commit } tl
+      | { maybe_packed = None; _ }, (commit, full_ref) :: tl when is_ref full_ref ->
+          go { acc with maybe_packed = Some commit } tl
+      | _, (_, full_ref) :: _ when is_non_packed_ref full_ref || is_ref full_ref ->
```

Comment:
Maybe this shouldn't fail if both refs point to the same commit.

Reply:
I think it should because if the ref is both a tag and branch, it might point to different commits in the future.

It would also need quite a bit of complex logic to deal with all possible cases with packed refs.

As I said above, I agree this isn't great but it wasn't any better before. I think the proper fix is to explicitly specify whether to track a branch or a tag, in which case this ambiguity doesn't exist anymore. I'm happy to do that in a separate PR as I believe this already fixes a bug in the current implementation and there's value in the PR as it is.

Reply:
We could improve the error message though and change it to `"Several refs share this name"` or something similar.

Reply:
Okay then I'm fine with that !

Reply:
I'll change the message, I think it's clearer!